File: problems/graphs/airport_connections/js/airport_connections.js
export const airportConnections = (() => {
  const initGraph = (airports, routes) => {
    const graph = {};
    for (const airport of airports) {
      graph[airport] = [];
    }

    for (const [from, to] of routes) {
      if (graph[from]) graph[from].push(to);
    }
    return graph;
  };

  const dfs = (graph, airport, visited, stack = null) => {
    visited.add(airport);

    for (const neighbor of graph[airport]) {
      if (!visited.has(neighbor)) {
        dfs(graph, neighbor, visited, stack);
      }
    }
    if (stack) stack.push(airport);
  };

  return (airports, routes, startingAirport) => {
    const graph = initGraph(airports, routes);

    const visited = new Set();
    const stack = [];
    for (const airport of airports) {
      if (!visited.has(airport)) {
        dfs(graph, airport, visited, stack);
      }
    }
    const reachable = new Set();
    dfs(graph, startingAirport, reachable);

    let newConnections = 0;
    while (stack.length) {
      const airport = stack.pop();

      if (!reachable.has(airport)) {
        newConnections++;
        dfs(graph, airport, reachable);
      }
    }
    return newConnections;
  };
})();

File: problems/graphs/airport_connections/cpp/airport_connections.cpp
#include "airport_connections.h"

void dfs(const AirportGraph &graph, const std::string &airport,
         std::unordered_set<std::string> &visited,
         std::stack<std::string> *order) {

  visited.insert(airport);

  if (graph.find(airport) != graph.end()) {
    for (const auto &neighbor : graph.at(airport)) {
      if (visited.find(neighbor) == visited.end()) {
        dfs(graph, neighbor, visited, order);
      }
    }
  }

  if (order)
    order->push(airport);
}

int airportConnections(const Airports &airports, const Routes &routes,
                       const std::string &startingAirport) {

  AirportGraph graph;
  for (const auto &airport : airports) {
    graph[airport] = {};
  }
  for (const auto &route : routes) {
    if (graph.find(route[0]) != graph.end())
      graph[route[0]].push_back(route[1]);
  }

  std::unordered_set<std::string> visited;
  std::stack<std::string> order;
  for (const auto &airport : airports) {
    if (visited.find(airport) == visited.end()) {
      dfs(graph, airport, visited, &order);
    }
  }

  visited.clear();
  dfs(graph, startingAirport, visited);

  int new_connections = 0;
  while (!order.empty()) {
    std::string airport = order.top();
    order.pop();

    if (visited.find(airport) == visited.end()) {
      new_connections++;
      dfs(graph, airport, visited);
    }
  }
  return new_connections;
}
